File: Source Problem/c++/프로그래머스/Level3/연습문제/N-Queen.py
answer = 0

def check(visited, cnt) :
    
    for i in range(0, cnt) :
        if visited[cnt] == visited[i] :
            return 0
        if cnt-i == abs(visited[cnt]-visited[i]) :
            return 0
    return 1

def dfs(visited, n, cnt) :
    
    if cnt == n:
        global answer
        answer += 1
        return
    
    for i in range(0, n) :
        visited[cnt] = i
        if check(visited, cnt) == 1 :
            dfs(visited, n, cnt+1)
    
    return

def solution(n):
    global answer
    visited = [0 for _ in range(n)]
    
    dfs(visited, n, 0)
 
    return answer

File: Source Problem/c++/프로그래머스/Level3/연습문제/N-Queen.cpp
#include <string>
#include <vector>

using namespace std;
int answer = 0;
int visited[16] = {0,};

bool check(int cnt){
    
    for(int i=0; i<cnt; i++){
        if(visited[cnt]==visited[i]){
            return false;
        }
        if(cnt-i == abs(visited[cnt]-visited[i])){
            return false;
        }
    }
    return true;
}

void dfs(int cnt, int target){
    
    if(cnt==target){
        answer ++;
        return;
    }
    
    for(int i=0; i<target; i++){
        visited[cnt] = i;
        if(check(cnt)==true){
            dfs(cnt+1, target);
        }
    }
    return;
}

int solution(int n) {
    dfs(0, n);
    return answer;
}